[AArch64][TargetParser] move CPUInfo into tablegen [NFC]

tommat01 opened the pull request:
This builds on top of https://github.com/llvm/llvm-project/pull/92037, which moves the architecture information. The new changes start at 52b8af6cde23d6ceacbc33f292e0bff1265501b5.

The changes to `AArch64Processors.td` look like a lot, but actually all that is happening for each processor is:
- The class is changed from `ProcessorModel` to `AArch64Processor`
- The base architecture is added, as a reference to the `Architecture64` e.g. `HasV9_0aOps`
- The list of "enabled by default" extensions is added.

tommat01 committed:
[AArch64][TargetParser] move ArchInfo into tablegen

tommat01 committed:
[AArch64][TargetParser] move CPUInfo into tablegen

tommat01 committed:
Remove "generic" from CPUInfo by making it a ProcessorModel

Making generic a AArch64Processor caused two test failures:
 - aarch64-targetattr.c (minor fix)
 - aarch64-fmv-dependencies.c (all dependencies completely wrong)

tommat01 committed:
Handle "generic" like any other CPU

tommat01 committed:
Fix AppleA14 and Saphira architecture definitions

tommat01 committed:
fix tests

The order of CPUs has changed

labrinea commented on llvm/lib/Target/AArch64/AArch64Processors.td:
```diff
@@ -812,178 +812,270 @@ def ProcessorFeatures {
+                     [FeatureAES, FeatureSHA2, FeatureCRC]>;
+def : AArch64Processor<"cortex-a55", HasV8_2aOps, CortexA55Model, ProcessorFeatures.A55,
+                     [TuneA55],
+                     [FeatureAES, FeatureSHA2, FeatureFullFP16, FeatureDotProd, FeatureRCPC]>;
```
As I said in the other review, why do we need to differentiate [FeatureAES, FeatureSHA2, FeatureFullFP16, FeatureDotProd, FeatureRCPC] from ProcessorFeatures.A55? If they differ it seems like a (historical) bug.

jroelofs replied:
And for that matter, why do we need `HasV8_2Ops` here when it's already in `ProcessorFeatures.A55`?

tmatheson-arm replied:
The answer having two separate lists for the Architecture definitions is that one is a mandatory list and the other is the "enabled by default" list. I kept the same structure here but I am less confident that it is needed, I will double check and merge the two lists if possible.

`HasV8_2Ops` is added so that there is a top-level reference to the architecture for this processor. In `ProcessorFeatures.A55` it serves to add all of the implied features of `HasV8_2Ops` to the processor. It is true there is some redundancy here, but these are the options I see:
- Keep this redundancy, and add a tablegen check to make sure they are always consistent.
- Use the first feature in the feature list as the architecture for this processor (and assert that the first feature is an architecture etc in tablegen). This removes the redundancy but it seems a bit "magic".
- Omit `HasV8_2Ops` from `ProcessorFeatures.A55` and then somehow programatically add the architecture and its implied features when expanding the features for the cpu. I can look into this but it seemed like it was going to be more obscure.

Do you have a preference or another suggestions?

jroelofs replied:
This looks much better.

jroelofs replied:
I still don't follow why `HasV8_2Ops` needs to be an entry on `AArch64Processor` instead of it always living on the `ProcessorFeatures` record. What do we gain from having it there?

jroelofs replied:
It does sound a little better, though I must admit I am partially motivated by having downstream diff in this area. Churn here, and in TargetParserTest.cpp can be kinda painful, so I want to make sure that pain is for a good cause.

jroelofs commented on llvm/utils/TableGen/ARMTargetDefEmitter.cpp:
```diff
@@ -108,6 +111,101 @@ static void EmitARMTargetDef(RecordKeeper &RK, raw_ostream &OS) {
+    return Minor == 0 ? "ARMV" + std::to_string(Major) + Profile.upper()
+                      : "ARMV" + std::to_string(Major) + "_" +
+                            std::to_string(Minor) + Profile.upper();
+  };
```
I think this lambda's return type is `Twine`? Probably need `.str()`s here so it's not returning dangling references to temporaries.

jroelofs replied:
oh, I see why now. anyway, thanks!

tommat01 committed:
make lambda return type explicit

tommat01 committed:
fix tests to account for new "generic" cpu, bump number of CPUs

tommat01 committed:
Replace FeatureCrypto with AES+SHA2 for all CPUs

This avoids the need to change a lot of TargetParserTest cases

tommat01 committed:
combine Implies and DefaultExts for cpus

Fold CPU default extensions into the list of SubtargetFeatures

TargetParserTest fixes because Implies and DefaultExts are combined

Now that they are combined, more things are accounted for in the
targetparser, i.e. more extensions are listed as being enabled for the
given CPUs.

clang test fixes because Implies and DefaultExts are combined

tommat01 committed:
Merge remote-tracking branch 'upstream/main' into targetparser_from_tablegen_cpuinfo

tommat01 committed:
remove unnecessary AEK_PERFMON additions

jroelofs commented on llvm/lib/Target/AArch64/AArch64Processors.td:
```diff
@@ -618,187 +618,210 @@ def TuneAmpere1B : SubtargetFeature<"ampere1b", "ARMProcFamily", "Ampere1B",
                                      FeatureNEON, FeaturePerfMon, FeatureFullFP16,
                                      FeatureFP16FML, FeatureSHA3];
-  list<SubtargetFeature> AppleA14 = [HasV8_4aOps, FeatureCrypto, FeatureFPARMv8,
+  list<SubtargetFeature> AppleA14 = [HasV8_5aOps, FeatureSHA2, FeatureAES, FeatureFPARMv8,
```
The `v8.4a => v8.5a` change looks correct, but mind landing that in a separate patch first?

tmatheson-arm replied:
https://github.com/llvm/llvm-project/pull/92600

tommat01 committed:
Remove AArch64Processor class

The only thing this added over ProcModel was the Architecture field.
Instead, search the list of Features to find the architecture for
each processor.

Since the "generic" processor does not have an Architecture, it needs
special treatment in the emitter.

jroelofs commented on llvm/lib/Target/AArch64/AArch64Processors.td:
```diff
@@ -812,178 +835,192 @@ def ProcessorFeatures {
+def : AArch64Processor<"apple-a15", HasV8_6aOps, CycloneModel, ProcessorFeatures.AppleA15,
                      [TuneAppleA15]>;
-def : ProcessorModel<"apple-a16", CycloneModel, ProcessorFeatures.AppleA16,
+// Alias for the latest Apple processor model supported by LLVM.
```
Let's leave this copy of the comment out. It will have to move every time this changes, which is not great for auto-merging.

tommat01 committed:
Remove extra comment about apple-latest alias

tommat01 committed:
clang-format

tommat01 committed:
remove whitespace only changes

tommat01 committed:
Merge remote-tracking branch 'upstream/main' into targetparser_from_tablegen_cpuinfo

tommat01 committed:
fix clang/test/Misc/target-invalid-cpu-note.c

tommat01 committed:
Merge remote-tracking branch 'upstream/main' into targetparser_from_tablegen_cpuinfo

jthackray reviewed (approved):
LGTM, assuming it causes no regressions.

tommat01 committed:
Merge remote-tracking branch 'upstream/main' into targetparser_from_tablegen_cpuinfo

tommat01 committed:
Merge remote-tracking branch 'upstream/main' into targetparser_from_tablegen_cpuinfo

tommat01 committed:
Merge remote-tracking branch 'upstream/main' into targetparser_from_tablegen_cpuinfo

pratlucas reviewed (approved):
LGTM.

tommat01 committed:
Move commit

tommat01 committed:
Fix TargetParserTest

tommat01 committed:
clang-format

tommat01 committed:
Merge remote-tracking branch 'upstream/main' into targetparser_from_tablegen_cpuinfo
